### ISSUE TITLE
HeaderDecoder IndexOutOfBoundsException

### ISSUE THREAD

**Issue by dhleemarchex** [nrinaudo/kantan.csv#179]
```
import kantan.csv._
import kantan.csv.ops._

case class Foo(header1: String, header2: Int)

implicit val decoder: HeaderDecoder[Foo] = HeaderDecoder.decoder("head_1", "head_2")(Foo.apply)

val csv1 = "head_2\n3"
csv1.readCsv[List, Foo](rfc.withHeader)

val csv2 = "head_1\nvalue1"
csv2.readCsv[List, Foo](rfc.withHeader)
```

returns
```
csv1: String = head_2
3
res0: List[kantan.csv.ReadResult[Foo]] = List(Left(TypeError: Missing header for head_2))

csv2: String = head_1
value1
java.lang.IndexOutOfBoundsException: 1
	at scala.collection.mutable.ResizableArray.apply(scratch_2.scala:42)
	at scala.collection.mutable.ResizableArray.apply$(scratch_2.scala:41)
	at scala.collection.mutable.ArrayBuffer.apply(scratch_2.scala:45)
	at kantan.csv.HeaderDecoder$.decoderFromMap(scratch_2.scala:46)
	at kantan.csv.GeneratedHeaderDecoders$$anon$2.fromHeader(scratch_2.scala:53)
	at kantan.csv.CsvReader$.$anonfun$apply$2(scratch_2.scala:34)
	at scala.util.Either$RightProjection.flatMap(scratch_2.scala:697)
	at kantan.csv.CsvReader$.apply(scratch_2.scala:34)
	at kantan.csv.CsvSource.$anonfun$reader$1(scratch_2.scala:62)
	at scala.util.Either$RightProjection.map(scratch_2.scala:709)
	at kantan.csv.CsvSource.reader(scratch_2.scala:62)
	at kantan.csv.CsvSource.reader$(scratch_2.scala:60)
	at kantan.csv.CsvSource$$anon$1.reader(scratch_2.scala:228)
	at kantan.csv.CsvSource.read(scratch_2.scala:124)
	at kantan.csv.CsvSource.read$(scratch_2.scala:120)
	at kantan.csv.CsvSource$$anon$1.read(scratch_2.scala:228)
	at kantan.csv.ops.CsvSourceOps.readCsv(scratch_2.scala:126)
	at #worksheet#.get$$instance$$res1(scratch_2.scala:12)
	at #worksheet#.#worksheet#(scratch_2.scala:35)
```

Looks like the issue is here https://github.com/nrinaudo/kantan.csv/blob/master/core/shared/src/main/scala/kantan/csv/HeaderDecoder.scala#L50
where it indexes into header instead of map.keys

**Comment by dhleemarchex** [comment]
The first test case might be related to https://github.com/nrinaudo/kantan.csv/issues/162

**Comment by nrinaudo** [comment]
Hey thanks! I appreciate the time you took to report this properly and even think of a work around.

I'll try to give this the attention it deserves as soon as possible. Might not be until next week though, I hope this is not a critical issue for you?

**Comment by dhleemarchex** [comment]
It's not critical, we're currently using a custom header decoder with the fix inside

**Comment by nrinaudo** [comment]
Pretty sure you're right and that your PR also fixes #162. Thank you very much!

**Comment by nrinaudo** [comment]
Fixed, thanks to a @dhleemarchex PR that I completely messed up when merging it in. The good bits are his, whatever is broken is me failing at git.
